File: python/longest_string_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        max_length = 0
        left = 0
        right = 1

        while left < len(s):
            left_char = s[left]
            char_set.add(left_char)

            if right >= len(s):
                max_length = max_length or 1
                break

            right_char = s[right]
            while right_char not in char_set:
                char_set.add(right_char)

                if right + 1 < len(s):
                    right += 1
                    right_char = s[right]

                this_len = len(char_set)
                if this_len > max_length:
                    max_length = this_len

            char_set.remove(left_char)
            left += 1
            if left >= right:
                right = left + 1

        return max_length


if __name__ == "__main__":
    print(Solution().lengthOfLongestSubstring("dvdf"))

File: src/main/java/com/leet/modules/LongestStringWithoutRepeatingCharacters.java
package com.leet.modules;

import com.leet.interfaces.Solution;

import java.util.HashSet;

public class LongestStringWithoutRepeatingCharacters implements Solution {
  public String solve(String[] args) {
    String s = args[0];
    Integer maxLength = 0;
    Integer left = 0;
    Integer right = 1;
    HashSet<Character> charSet = new HashSet<Character>();

    while (left < s.length()) {
      Character leftChar = s.charAt(left);
      System.out.println("CharSet: " + charSet);

      if (!charSet.contains(leftChar)) {
        charSet.add(leftChar);
        System.out.println("Adding Left: " + leftChar);
      }

      if (right >= s.length()) {
        maxLength = (maxLength == 0) ? 1 : maxLength;
        break;
      }

      Character rightChar = s.charAt(right);
      while (!charSet.contains(rightChar)) {
        charSet.add(rightChar);
        System.out.println("Adding Right: " + rightChar);

        if (right + 1 < s.length()) {
          right += 1;
          rightChar = s.charAt(right);
        }
        if (charSet.size() > maxLength) {
          maxLength = charSet.size();
          System.out.println("New Max Size: " + maxLength);
        }
      }

      System.out.println("Final CharSet: " + charSet);
      System.out.println("=".repeat(20));

      charSet.remove(leftChar);
      left += 1;
      if (right <= left) {
        right += 1;
      }
    }
    return maxLength.toString();
  };
}
